File: Lodjinha/app/src/main/java/freitas/dias/lodjinha/adapter/CategoriesAdapter.java
package freitas.dias.lodjinha.adapter;


import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.List;

import freitas.dias.lodjinha.R;
import freitas.dias.lodjinha.api.model.Category;

public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.ViewHolder>{

    private List<Category> categories;
    private Context context;

    public CategoriesAdapter(Context context, List<Category> categories) {
        this.categories = categories;
        this.context = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_categories, null);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Category category = categories.get(position);

        if (!TextUtils.isEmpty(category.getUrlImage())) {
            Picasso.with(context).load(category.getUrlImage())
                    .into(holder.imageView);
        }

        holder.textView.setText(category.getDescription());
    }

    @Override
    public int getItemCount() {
        return categories.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ImageView imageView;
        TextView textView;

        ViewHolder(View view) {
            super(view);
            this.imageView = (ImageView) view.findViewById(R.id.recyclerView_categories_image);
            this.textView = (TextView) view.findViewById(R.id.recyclerView_categories_name);
        }
    }
}

File: app/src/main/java/com/example/lidjinha/lodjinha/ui/home/CategoriesAdapter.kt
package com.example.lidjinha.lodjinha.ui.home

import android.content.Context
import android.content.Intent
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.example.lidjinha.lodjinha.R
import com.example.lidjinha.lodjinha.model.Categorie
import com.example.lidjinha.lodjinha.ui.productscategory.ProductsCategoryActivity
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

class CategoriesAdapter(private val categories: List<Categorie>,
                        private val context: Context) :
        RecyclerView.Adapter<CategoriesAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_categories_list, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        Picasso.with(context).load(categories[position].urlImage).into(holder.image, callbackImage(holder.image))
        holder.description.text = categories[position].description
        holder.itemView.setOnClickListener {
            val intent = Intent(context, ProductsCategoryActivity::class.java)
            intent.putExtra(ProductsCategoryActivity.EXTRA_CATEGORIE, categories[position])
            context.startActivity(intent)
        }
    }

    private fun callbackImage(image: ImageView) = object : Callback {
        override fun onSuccess() {
            Log.d("Load categories", "Success!")
        }

        override fun onError() {
            image.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_down_black))
        }
    }

    override fun getItemCount(): Int {
        return categories.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val image: ImageView = itemView.findViewById(R.id.iv_categorie_image)
        val description: TextView = itemView.findViewById(R.id.tv_categorie_description)

    }
}